minitask3/akhmad/follower.py

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python
+import rospy
+import cv2
+import cv_bridge
+import numpy
+
+from geometry_msgs.msg import Twist
+from sensor_msgs.msg import Image, LaserScan
+
+LOWER_GREEN = numpy.array([60, 100, 50])
+UPPER_GREEN = numpy.array([60, 255, 255])
+
+SAMPLE_WINDOW = 20  # Degrees
+
+TURN_MAX_SPEED = 0.2
+TURN_GAIN = 0.01
+
+FORWARD_SPEED = 0.3
+
+
+class Follower():
+
+    def __init__(self):
+        rospy.init_node('robot_control', anonymous=True)
+        self.vel_pub = rospy.Publisher("/cmd_vel", Twist, queue_size=1)
+        self.bridge = cv_bridge.CvBridge()
+
+        self.move_twist = Twist()
+
+        self.front = False
+        self.right = False
+        self.green_block_detected = False
+        self.centroids = []
+        self.width_middle = None
+
+    def start(self):
+        self.scan_sub = rospy.Subscriber("/scan", LaserScan, self.onscan)
+        self.camera_sub = rospy.Subscriber("/camera/rgb/image_raw",
+                                           Image, self.onimage)
+        rospy.spin()
+
+    def move(self, turnspeed, forwardspeed):
+        '''
+        Move the robot forward at forwardspeed, and turn at turnspeed.
+        '''
+        self.move_twist.linear.x = forwardspeed
+        self.move_twist.angular.z = turnspeed
+        self.vel_pub.publish(self.move_twist)
+
+    def sample_scan(self, scandata, target, window_size=SAMPLE_WINDOW):
+        '''
+        Returns the minimum value in the given window_size, centered on target
+        '''
+        assert type(window_size) == int, "window_size should be an integer"
+        offset = window_size / 2
+
+        min_val = scandata.ranges[target - offset]
+        for i in range(target - offset + 1, target + offset):
+            min_val = min(scandata.ranges[i], min_val)
+        return min_val
+
+    def move_by_image(self):
+        rospy.loginfo("Walking towards green block")
+        err_x = float('inf')
+        err_y = float('inf')
+
+        # Walk towards block that is closest to middle of image width
+        for c in self.centroids:
+            err_x = min(c['cx'] - self.width_middle, err_x)
+            err_y = min(c['cy'] - self.height_middle, err_y)
+        turnspeed = -float(err_x) * TURN_GAIN
+
+        # Make sure turnspeed is not too high
+        turnspeed = TURN_MAX_SPEED if turnspeed > TURN_MAX_SPEED else turnspeed
+        turnspeed = -TURN_MAX_SPEED if turnspeed < -TURN_MAX_SPEED else turnspeed
+
+        # Was trying to stop robot in front of block but this doesn't work
+        if (err_y > -10 and err_y < 10 and self.front):
+            self.move(0, 0)
+        else:
+            self.move(turnspeed, FORWARD_SPEED)
```

Comment:
There is no explicit requirement to do this. It is definitely a nice idea but if it is not working it may be best to comment it out or remove it entirely

Reply:
I've tweaked this block of code. Now it is working most of the times